## Introduction:

This section in the textbook is about the relative percentile rank .. 

### Ex:
I recently ran the James Joyce Ramble 10K in Dedham MA. The results are available from http://coolrunning.com/results/10/ma/Apr25_27thAn_set1.shtml.

1. Go to that page and find my results. I came in 97th in a field of 1633, so what is my percentile rank in the field?
   In my division (M4049 means “male between 40 and 49 years of age”) I came in 26th out of 256. 
   What is my percentile   rank in my division? 
   
2. If I am still running in 10 years (and I hope I am), I will be in the M5059 division. Assuming that my percentile rank in my division is the same, how much slower should I expect to be? 

3. I maintain a friendly rivalry with a student who is in the F2039 division. How fast does she have to run her next 10K to “beat” me in terms of percentile ranks?

In [4]:
using DataFrames
using Plots

In [5]:
df = readtable("results.csv", header = false, names=[:Place, :DivTotal, :Div, :Guntime, :Nettime, :Pace])

1596×6 DataFrames.DataFrame
│ Row  │ Place │ DivTotal  │ Div     │ Guntime   │ Nettime   │ Pace    │
├──────┼───────┼───────────┼─────────┼───────────┼───────────┼─────────┤
│ 1    │ 1     │ "1/362"   │ "M2039" │ "30:43"   │ "30:42"   │ "4:57"  │
│ 2    │ 2     │ "2/362"   │ "M2039" │ "31:36"   │ "31:36"   │ "5:06"  │
│ 3    │ 3     │ "3/362"   │ "M2039" │ "31:42"   │ "31:42"   │ "5:07"  │
│ 4    │ 4     │ "4/362"   │ "M2039" │ "32:28"   │ "32:27"   │ "5:14"  │
│ 5    │ 5     │ "5/362"   │ "M2039" │ "32:52"   │ "32:52"   │ "5:18"  │
│ 6    │ 6     │ "6/362"   │ "M2039" │ "33:06"   │ "33:06"   │ "5:20"  │
│ 7    │ 7     │ "1/256"   │ "M4049" │ "33:33"   │ "33:33"   │ "5:24"  │
│ 8    │ 8     │ "2/256"   │ "M4049" │ "33:42"   │ "33:42"   │ "5:26"  │
│ 9    │ 9     │ "3/256"   │ "M4049" │ "33:54"   │ "33:54"   │ "5:28"  │
│ 10   │ 10    │ "7/362"   │ "M2039" │ "33:56"   │ "33:56"   │ "5:28"  │
│ 11   │ 11    │ "8/362"   │ "M2039" │ "34:13"   │ "34:13"   │ "5:31"  │
⋮
│ 1585 │ 1622  │ "61/64"   │ "F5059" │ "1:30:35" │ "1:28:56" │ "14:19" │
│ 1586 │ 1623  │ "444/448" │ "F2039" │ "1:31:52" │ "1:30:13" │ "14:32" │
│ 1587 │ 1624  │ "445/448" │ "F2039" │ "1:30:19" │ "1:30:19" │ "14:33" │
│ 1588 │ 1625  │ "45/46"   │ "M6069" │ "1:32:31" │ "1:31:15" │ "14:41" │
│ 1589 │ 1626  │ "446/448" │ "F2039" │ "1:32:10" │ "1:32:01" │ "14:49" │
│ 1590 │ 1627  │ "447/448" │ "F2039" │ "1:34:13" │ "1:32:11" │ "14:50" │
│ 1591 │ 1628  │ "62/64"   │ "F5059" │ "1:34:26" │ "1:32:24" │ "14:53" │
│ 1592 │ 1629  │ "171/171" │ "M5059" │ "1:50:09" │ "1:50:09" │ "17:44" │
│ 1593 │ 1630  │ "63/64"   │ "F5059" │ "1:55:13" │ "1:53:38" │ "18:18" │
│ 1594 │ 1631  │ "64/64"   │ "F5059" │ "1:55:14" │ "1:53:38" │ "18:18" │
│ 1595 │ 1632  │ "46/46"   │ "M6069" │ "1:55:34" │ "1:55:34" │ "18:36" │
│ 1596 │ 1633  │ "448/448" │ "F2039" │ "2:03:26" │ "2:01:47" │ "19:36" │

In [6]:
pace = df[:][6]

1596-element DataArrays.DataArray{String,1}:
 "4:57" 
 "5:06" 
 "5:07" 
 "5:14" 
 "5:18" 
 "5:20" 
 "5:24" 
 "5:26" 
 "5:28" 
 "5:28" 
 "5:31" 
 "5:33" 
 "5:42" 
 ⋮      
 "14:19"
 "14:32"
 "14:33"
 "14:41"
 "14:49"
 "14:50"
 "14:53"
 "17:44"
 "18:18"
 "18:18"
 "18:36"
 "19:36"

In [7]:
function ConverPaceToSpeed(pace)
    
    minutes = zeros(length(pace), 1)
    seconds = zeros(length(pace), 1)
    
    ## splitting pace column into minutes and seconds
    for i = 1:length(pace)
        if(length(pace[i][:]) == 5)  # width of the row in a pace vector
            minutes[i] = parse(Int, pace[i][1:end-3]) 
            seconds[i] = parse(Int, pace[i][4:5])
        else 
            minutes[i] = parse(Int, pace[i][1:end-3])
            seconds[i] = parse(Int, pace[i][3:4])
          end
    end
    
    seconds = minutes*60 + seconds
    speed = [1/i*60*60 for i in seconds]
    
    return speed
end

ConverPaceToSpeed (generic function with 1 method)

In [8]:
speed = ConverPaceToSpeed(pace)

1596×1 Array{Float64,2}:
 12.1212 
 11.7647 
 11.7264 
 11.465  
 11.3208 
 11.25   
 11.1111 
 11.0429 
 10.9756 
 10.9756 
 10.8761 
 10.8108 
 10.5263 
  ⋮      
  4.19092
  4.12844
  4.12371
  4.08627
  4.04949
  4.04494
  4.03135
  3.38346
  3.27869
  3.27869
  3.22581
  3.06122

In [9]:
c = countmap(speed) 

Dict{Float64,Int64} with 374 entries:
  9.52381 => 1
  8.37209 => 2
  7.94702 => 4
  6.09137 => 7
  5.88235 => 2
  6.29371 => 7
  7.27273 => 7
  9.89011 => 1
  9.13706 => 1
  6.79245 => 9
  6.90979 => 5
  5.74163 => 7
  5.7971  => 6
  4.99307 => 2
  8.07175 => 2
  6.71642 => 6
  5.84416 => 2
  5.21739 => 1
  5.07757 => 2
  7.79221 => 9
  7.48441 => 7
  6.02007 => 4
  7.69231 => 1
  6.48649 => 4
  6.07083 => 6
  ⋮       => ⋮

In [32]:
## calc cdf
function calc_cdf(c) # c is a data dict
    runsum = 0
    xs = Float64[]
    cs = Float64[]

    for (value, count) in sort(c)
        runsum = runsum + count
        push!(xs, value)
        push!(cs, runsum)
    end

    total = float(runsum)
    cdf = [c/total for c in cs]   ## cdf[end] = 1, ps carrried cdf
    
    return xs, cdf
end

calc_cdf (generic function with 1 method)

In [33]:
xs, cdf = calc_cdf(c)

([3.06122,3.22581,3.27869,3.38346,4.03135,4.04494,4.04949,4.08627,4.12371,4.12844  …  10.8761,10.9756,11.0429,11.1111,11.25,11.3208,11.465,11.7264,11.7647,12.1212],[0.000626566,0.00125313,0.00250627,0.00313283,0.0037594,0.00438596,0.00501253,0.0056391,0.00626566,0.00689223  …  0.993734,0.994987,0.995614,0.996241,0.996867,0.997494,0.99812,0.998747,0.999373,1.0])

In [34]:
plot(xs, cdf)
xlabel!("Values")
ylabel!("CDFs")

In [35]:
## let's calc the percentile rank .. 

function Percentile_Rank(scores, your_score)
    count = 0
    
    for score in scores
        if score <= your_score
            count += 1
        end
    end

    percentile_rank = 100.0 * count / length(scores)
    
    return percentile_rank
end

Percentile_Rank (generic function with 1 method)

In [36]:
author_speed = speed[end] 

3.0612244897959187

In [37]:
Percentile_Rank(speed, author_speed)

0.06265664160401002

Matches Python results .. 

and it is a reasonable result .. as the rank of the author in the race (the last) .. 

hence, all the competitors were better then him .. 

this is why his percentile rank was very tiny .. 

To assure this way of thinking .. let's calc the percentile rank of the winner (speeds[0]) ..  that is expected to be 100% .. which means that he beats 100% of the participants .. 

In [38]:
perc_rank = Percentile_Rank(speed, speed[1])
perc_rank

100.0

As expected .. !

## 1. Division of M4049:

In [41]:
M4049 = String[]

for i = 1:size(df)[1]
    if(df[3][i] == "M4049")  ## df is transposed
        push!(M4049, df[6][i])  # list of speeds within the input age group(not-converted)
    end
end
M4049

256-element Array{String,1}:
 "5:24" 
 "5:26" 
 "5:28" 
 "5:33" 
 "5:42" 
 "6:08" 
 "6:09" 
 "6:12" 
 "6:13" 
 "6:19" 
 "6:21" 
 "6:24" 
 "6:25" 
 ⋮      
 "11:24"
 "11:25"
 "11:28"
 "11:29"
 "11:37"
 "11:49"
 "11:55"
 "11:56"
 "12:16"
 "12:39"
 "12:41"
 "13:11"

In [42]:
length(M4049)  ## matches python

256

In [43]:
speeds_M4049 = ConverPaceToSpeed(M4049)

256×1 Array{Float64,2}:
 11.1111 
 11.0429 
 10.9756 
 10.8108 
 10.5263 
  9.78261
  9.7561 
  9.67742
  9.65147
  9.49868
  9.44882
  9.375  
  9.35065
  ⋮      
  5.26316
  5.25547
  5.23256
  5.22496
  5.16499
  5.07757
  5.03497
  5.02793
  4.8913 
  4.74308
  4.73062
  4.5512 

In [44]:
speeds_M4049[27] ## the author's speed ..

8.591885441527447

Note that in python we accessed the index of 26 not 27 .. cuase julia is a one based programming lang. and python is a zero based ..

In [45]:
perc_rank_M4049 = Percentile_Rank(speeds_M4049, speeds_M4049[27])
perc_rank_M4049

89.84375

Matches python .. 

His percetile rank in his division is about 89.8% .. 

## 2. Division of M5059:

Assuming that my percentile rank in my division is the same, how much slower should I expect to be? ..

Hence, I gonna calc. the actual percentile rank for him with respect to his new age group .. and then subtract the given percentile rank from ..

In [47]:
M5059 = String[]

for i = 1:size(df)[1]
    if(df[3][i] == "M5059")  ## df is transposed
        push!(M5059, df[6][i])  # list of speeds within the input age group(not-converted)
    end
end
M5059

171-element Array{String,1}:
 "6:11" 
 "6:17" 
 "6:24" 
 "6:39" 
 "6:42" 
 "6:44" 
 "6:49" 
 "6:58" 
 "7:04" 
 "7:11" 
 "7:12" 
 "7:16" 
 "7:18" 
 ⋮      
 "11:25"
 "11:32"
 "11:36"
 "11:38"
 "11:40"
 "11:43"
 "12:02"
 "12:03"
 "12:18"
 "12:22"
 "12:34"
 "17:44"

In [48]:
speeds_M5059 = ConverPaceToSpeed(M5059)

171×1 Array{Float64,2}:
 9.7035 
 9.54907
 9.375  
 9.02256
 8.95522
 8.91089
 8.80196
 8.61244
 8.49057
 8.35267
 8.33333
 8.25688
 8.21918
 ⋮      
 5.25547
 5.20231
 5.17241
 5.15759
 5.14286
 5.12091
 4.98615
 4.97925
 4.87805
 4.85175
 4.77454
 3.38346

In [49]:
speeds_M5059[27]  ## same speed position .. 

7.843137254901961

In [50]:
perc_rank_M5059 = Percentile_Rank(speeds_M5059, speeds_M5059[27])
perc_rank_M5059

85.38011695906432

Hence, difference is .. 

In [51]:
perc_rank_M4049 - perc_rank_M5059

4.463633040935676

## Conclusion:

His percentile rank is expected to be about 4.46% less than the given speed .. given his age group of 50-59 

## 3. Division of F2039:

I maintain a friendly rivalry with a student who is in the F2039 division. 

How fast does she have to run her next 10K to “beat” me in terms of percentile ranks?

In [52]:
F2039 = String[]

for i = 1:size(df)[1]
    if(df[3][i] == "F2039")  ## df is transposed
        push!(F2039, df[6][i])  # list of speeds within the input age group(not-converted)
    end
end
F2039

448-element Array{String,1}:
 "5:53" 
 "5:56" 
 "6:00" 
 "6:05" 
 "6:12" 
 "6:22" 
 "6:35" 
 "6:36" 
 "6:39" 
 "6:54" 
 "6:55" 
 "6:56" 
 "7:01" 
 ⋮      
 "12:01"
 "12:04"
 "12:13"
 "12:17"
 "12:28"
 "12:42"
 "13:11"
 "14:32"
 "14:33"
 "14:49"
 "14:50"
 "19:36"

In [53]:
speeds_F2039 = ConverPaceToSpeed(F2039)

448×1 Array{Float64,2}:
 10.1983 
 10.1124 
 10.0    
  9.86301
  9.67742
  9.42408
  9.11392
  9.09091
  9.02256
  8.69565
  8.6747 
  8.65385
  8.55107
  ⋮      
  4.99307
  4.97238
  4.91132
  4.88467
  4.81283
  4.72441
  4.5512 
  4.12844
  4.12371
  4.04949
  4.04494
  3.06122

Now, I need the speed of that girl .. the point is she is the least one in the whole race .. and consiquantly, she is the least one in her age group .. 

In [54]:
speeds_F2039[end]

3.0612244897959187

In [55]:
perc_rank_F2039 = Percentile_Rank(speeds_F2039, speeds_F2039[end])
perc_rank_F2039

0.22321428571428573

Her percentile rank is tiny (0.22) .. and this is expected as she ranks the last ! 

In [56]:
## to calc the rate .. i have to divide the speeds .. 
speeds_M5059[26] / speeds_F2039[end]

2.5620915032679736

This means that he is faster than her double and half .. per 1 Km ..

She has to enhance her speed by double and half to be able to beat him.